fix: misc builder fixes for builder tool and public forms, add MarkdownText component

justynoh opened the pull request:
## Problem
1. #5058 
2. #5050 
3. ~~#5055~~
4. #5061 

## Solution

1. Add descriptions for toggles that mimic tooltips that exist in angular in `EditTable`.
2. Add overflow-wrap in `FormLabel`, add overflow hidden in `ThenShowBlock`, add wrap anywhere in checkbox and radio labels.
3. ~~Toasts for paragraph field are determined by the title of the field determined by the question title for Statement/Paragraph fields. Change this to paragraph in `EditParagraph`. I THINK this is backward compatible, since the name of the questions isn't enforced anywhere, and we filter out entire statements/paragraphs during submission.~~ It is backwards compatible but makes for strange UI (statements created in angular are called "Statement" while those created in react are called "Paragraph", so the toasts might show differently depending on actions). I'll open another PR to fix it with a more sweeping change.
4. This one was a bit complicated. Looking at the angular app, we realize that links are automatically created from URLs in many more places in the app than we accounted for:
  - Form instructions
  - Header descriptions
  - Paragraphs
  - Image captions
  - Form descriptions
  - Thank you page description
 
  Change all of these to make use of ReactMarkdown. But react markdown does a weird thing where it sucks up consecutive newlines... (see [this issue](https://github.com/remarkjs/react-markdown/issues/278)) To get newlines to show up, wrap them in a box with whitespace = pre-wrap. This means that **3 or more `\n`'s always looks like only 2 `\n`'s, because it gets rendered as two consecutive `<p>` tags by react markdown**. Thus add new processing to the `children` text to inject new-line nodes into the markdown AST.

**Breaking Changes** 
- No - this PR is backwards compatible

justynoh committed:
fix: add description for email confirmations, change paragraph names, use overflow wrap to cut off long words in form labels, form descriptions and checkbox and label options

justynoh committed:
feat: use ReactMarkdown where needed to emulate angular, modifications to wrap text so that it looks correct in logic blocks and form label titles

justynoh committed:
chore: remove mistaken mt from formlabel description

karrui commented on frontend/src/components/FormControl/FormLabel/FormLabel.tsx:
```diff
@@ -145,7 +148,12 @@ const FormLabelDescription = ({
-    <ReactMarkdown components={mdComponents}>{children}</ReactMarkdown>
+    // Wrap markdown with a <div white-space='pre-wrap'> to get consecutive newlines to show up
+    <Box whiteSpace="pre-wrap">
+      <ReactMarkdown components={mdComponents} remarkPlugins={[gfm]}>
```
as discussed, should we create a separate `TextMarkdown` component that uses all the plugins we want to have a consistent markdown usage across the app?

also use something like [remark-breaks](https://github.com/remarkjs/remark-breaks) so breaks are (in)correctly rendered. (breaks markdown spec but for our use case its fine lol)

justynoh replied:
looking at https://github.com/vmware-tanzu/kubeapps/pull/4765 but it didn't work - some error with gfm, breaks and react-markdown not playing nice together.

Tried following this "little plugin" but it's still not working (now it shows up worse than it would have with pre-wrap): https://github.com/remarkjs/react-markdown/issues/278#issuecomment-971895356

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/25571626/195064054-1b961bd3-9cf6-4a5a-a582-1845f9702630.png">

wanlingt reviewed (approved):
lgtm! (other than kr's comment)

nit: btw the react markdown link leads to the homepage and not the issue

justynoh committed:
feat: extract markdown to it's own component

karrui committed:
chore: upgrade react-markdown, add remarks-breaks package

for creating appropriate new lines in app

karrui committed:
feat: update MarkdownText component to generate correct new lines

karrui committed:
feat: replace pre-line CSS to pre-wrap

so whitespace is not collapsed, and should be closer to original intent. Mainly affects Modal, drawers, etc due to copy paste

karrui commented:
ok hope stories catches all regressions lol

karrui committed:
feat: add whiteSpace=pre-wrap useMdComponents styles

for consistency

justynoh committed:
chore: revert changing of paragraph title due to issues with existing fields containing title Statement being confusing for users.

karrui committed:
feat: only allow multiline breaks in editable text

so the WYSIWYG behaviour only stays in editable field's outputs instead of ruining entirety of markdown spec for other uses.

karrui committed:
fix(Banner): update banner story text to not have new line

karrui committed:
feat(EditAttachment): correct whitespace in markdown text

karrui committed:
fix(story): fix whitespace rendering for instructions in form stories

karrui commented on frontend/src/components/MarkdownText/MarkdownText.tsx:
```diff
@@ -6,18 +6,27 @@ import gfm from 'remark-gfm'
 interface MarkdownTextProps {
   components?: Components
+  /**
+   * Whether to allow sequetial new lines to generate sequential line breaks.
```
lazy to push again to fix this typo...

karrui replied:
[9a69184](https://github.com/opengovsg/FormSG/pull/5105/commits/9a69184d9a06891a4341ac48b474fadffb0cd5a4)

karrui committed:
fix: typo in param doc

karrui commented on frontend/src/templates/Field/Paragraph/ParagraphField.tsx:
```diff
@@ -24,9 +25,12 @@ export const ParagraphField = ({
-    <ReactMarkdown components={mdComponents}>
-      {schema.description}
-    </ReactMarkdown>
+    <Box mt="1rem">
```
any reason why you added a margin top to this? @justynoh

justynoh replied:
shoot, no, no reason. should be deleted, i must have missed deleting it after copy paste oops

justynoh replied:
[7cbbfff](https://github.com/opengovsg/FormSG/pull/5105/commits/7cbbfff14b8b887817d5f4d8ae55d8b05023d9b3)

justynoh committed:
chore: remove margin top from paragraph field
